### ISSUE TITLE
False positive S4325: Redundant casts

### ISSUE THREAD

**Issue by ajmath** [SonarSource/SonarTS#828]
**I want to report a bug.**

**SonarTS version:** I'm not sure where this is supposed be found, is this a node module I'm suppsoed to include in my project?  Is it something that is supposed to be present at scan time?  Is it something that is configured on the server?
**Node.js version (local):** 8.10
**Node.js version (jenkins):** ? The scan is being run as a part of my companies CI pipeline and I don't have any control over that environment so I'm not sure what node version is running.
**TypeScript version:** 3.2.1

**SonarQube version:** 7.4 Developer edition build 18908
**SonarLint version and used IDE:** <br>
**TSLint version:** 5.11.0

**Rule key:** typescript:S4325

**Reproducer** 
```
interface RNPushNotification {
  getData: () => {[s: string]: string};
}
interface PushNotification {
  data: {[s: string]: string};
}
function isRnPushNotification(raw: RNPushNotification | PushNotification): raw is RNPushNotification {
  return typeof (raw as any).getData === "function";
}
```

**Logs**
Here is what the errors presented in SonarQube look like:
<img width="823" alt="image" src="https://user-images.githubusercontent.com/266679/51689279-587aba80-1fc4-11e9-9187-f417dc797c00.png">

And here is what Webstorm tells me when I try to remove the cast:
![image](https://user-images.githubusercontent.com/266679/51689337-79431000-1fc4-11e9-8e85-33184eb2e9ff.png)

**Expected behavior**
This should not be reported as an error or code smell

I'm seeing similar issues with this rule key as the one reported in #779  some react component props that I can share later once I have a chance to extract the relevant bits.

**Comment by ajmath** [comment]
tsconfig:
```json
{
  "compilerOptions": {
    "target": "es2017",
    "jsx": "react-native",
    "noEmit": false,
    "moduleResolution": "node",
    "module": "commonjs",
    "strict": true,
    "noUnusedLocals": true,
    "skipLibCheck": true,
    "forceConsistentCasingInFileNames": true,
    "noImplicitReturns": true,
    "allowSyntheticDefaultImports": true,
    "esModuleInterop": true,
    "resolveJsonModule": true,
    "allowJs": true,
    "checkJs": false,
    "outDir": ".tmp",
    "typeRoots": [
      "node_modules/@types",
      "app/types"
    ],
    "types":[],
    "plugins": [
      {
        "name": "tslint-language-service",
        "ignoreDefinitionFiles": true
      }
    ],
    "importHelpers": true,
    "baseUrl": ".",
    "paths": {
      "native-base": ["app/types/native-base/index.d.ts"]
    }
  },
  "include" : [
    "app/**/*",
    "__tests__/**/*"
  ],
  "exclude": [
    "node_modules",
    ".tmp",
    "app/native-base-theme"
  ]
}
```

**Comment by sylvanaar** [comment]
I have a very similar issue. I need a particular boolean type to be narrowed so I cast it, and the cast is flagged as unnecessary.

```typescript
const right = {
    error: true as true // cast is flagged as unnecessary
}

typeof right.error // true

const wrong = {
    error: true
}

typeof wrong.error; // boolean
```

**Comment by vilchik-elena** [comment]
@sylvanaar I created a ticket for your problem

**Comment by vilchik-elena** [comment]
@ajmath I can not reproduce with your code sample. Are you sure it is sufficient? 

The rule logic is the following. We will report on `(X as T).FOO` when you access a field `FOO` on a casted expression `X` which initial type is union where each member of this union has `FOO`.

In your example `getData` is available on `RNPushNotification` but not on `PushNotification`

**Comment by ajmath** [comment]
@vilchik-elena I'm getting some mixed results on my end as well.  When I add new code that's identical to the failing code shown above, it passes the sonar scanner without issue.  However, the same code from the initial scan is what is being reported as an error.

I'm digging into the differences in how those are executed on my end now.

Can you provide any info on how SonarTS integration with the scanner and the SonarQube server?  Specifically, how can I find out which version of SonarTS we're using?  I'm not an admin so it's possible this is out of my reach.

**Comment by ajmath** [comment]
I found something else interesting, while I didn't see an issue with the sample code above, copying the original erroring code into a new location caused the issue to pop up again.  The key difference between the two is that `RNPushNotification` and `PushNotification` are defined in external node_modules that aren't installed at scan time.  Could this be the reason?

**Comment by vilchik-elena** [comment]
To be honest I'm lost in your explanations:) I prefer if you start from scratch here. 

You don't need to analyze node_modules to have type information available during the scan of main files (but node_modules should be installed). SonarTS uses TypeScript compiler, so it will just launch it for your tsconfig file.

**Comment by ajmath** [comment]
@vilchik-elena Sorry for any confusion and thanks for your help.  I'll start over:

The reproducer I created above wasn't an actual reproducer because the types were defined in-line.  In the problem that initiated this (screenshot), the types were defined in node modules so I'm guessing that Sonar is interpreting them as any. 

TLDR: False alarm, the real culprit was node_modules not being installed at scan time.
